servant-client/src/Servant/Client/Generic.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+module Servant.Client.Generic
+  ( ClientLike(..)
+  , genericMkClient
+  ) where
+
+import Generics.SOP   (Generic, I(..), NP(..), NS(Z), Rep, SOP(..), to)
+import Servant.API    ((:<|>)(..))
+import Servant.Client (ClientM)
+
+-- | This class allows us to match client structure with client functions
+-- produced with 'client' without explicit pattern-matching.
+--
+-- The client structure needs a 'Generics.SOP.Generic' instance.
+--
+-- Example:
+--
+-- type API
+--     = "foo" :> Capture "x" Int :> Get '[JSON] Int
+--  :<|> "bar" :> QueryParam "a" Char :> QueryParam "b" String :> Post '[JSON] [Int]
+--  :<|> Captre "nested" Int :> NestedAPI
+--
+-- type NestedAPI
+--     = Get '[JSON] String
+--  :<|> "baz" :> QueryParam "c" Char :> Post '[JSON] ()
+--
+-- data APIClient = APIClient
+--   { getFoo         :: Int -> ClientM Int
+--   , postBar        :: Maybe Char -> Maybe String -> ClientM [Int]
+--   , mkNestedClient :: Int -> NestedClient
+--   } deriving GHC.Generic
+--
+-- instance Generic.SOP.Generic APIClient
+-- instance (Client API ~ client) => ClientLike client APIClient
+--
+-- data NestedClient = NestedClient
+--  { getString ::  ClientM String
+--  , postBaz   :: Maybe Char -> ClientM ()
+--  } deriving GHC.Generic
+--
+-- instance Generic.SOP.Generic
+-- instance (Client NestedAPI ~ client) => ClientLike client NestedAPI
+--
+-- mkAPIClient :: APIClient
+-- mkAPIClient = mkClient (client (Proxy :: Proxy API))
+class ClientLike client custom where
+  mkClient :: client -> custom
```

Comment:
Well, GHC simply does not allow type family synonym instances (even with `TypeSynonymInstances` enabled). So we work around that with type equality constraint. Since in practice client structure (e.g. `NestedClient`) will only have one instance, I don't see this as a problem (only as an inconvenience).
I am not sure why that's not allowed, but I didn't investigate it.

We have also explored the possibility of having `ClientLike` parametrized with an API type rather than client type. But that leads to many more problems (e.g. you now have to know API type for every recursing `mkClient` call and also have separate instances for API combinators).